src/coreclr/src/jit/hwintrinsic.cpp

```diff
@@ -538,9 +546,16 @@ GenTree* Compiler::addRangeCheckIfNeeded(NamedIntrinsic intrinsic, GenTree* immO
         {
             index = gtCloneExpr(immOp);
         }
+
+        if (immLowerBoundIncl != 0)
+        {
+            index = gtNewOperNode(GT_SUB, TYP_INT, index, gtNewIconNode(immLowerBoundIncl, TYP_INT));
```

Comment:
So this will now do:
```csharp
immOp -= lowerBound;

if (immOp > adjustedUpperBound) {
    throw new ArgumentOutOfRangeException();
}

switch (immOp) {
    // ...
}
```

?

Reply:
Almost, the code added in the importer doesn't alter the value of immOp. In other words, this what is tested
```c#
if ((immOp - lowerBound)> (upperBound - lowerBound)) {
    throw new ArgumentOutOfRangeException();
}
```

And then in the codegen the JIT will generate
```c#
switch (immOp - lowerBound) {
case 0:
case 1:
// ...
case upperBound-1:
}
```

Reply:
Wouldn't that mean that the `TestImmOpZeroOrOne` case (https://github.com/dotnet/runtime/pull/36830#discussion_r428868360) could be handled by this?
That is, if `lowerBoundIncl` is `1` and `upperBoundExcl` is `3`, then we could have `1` or `2` as valid values. Subtracting the lower bound then gives us `0` or `1` which allows us to use `cbnz`.
In any case, it seems fine to me. We might just want a comment explaining that the case numbers are offset as compared to the actual inputs.

Reply:
Perhaps, but I am not sure if this is going to be useful though. I haven't seen any instruction that has such ranges of its immediate operand.

Reply:
I will add the comment